relayer/src/foreign_client.rs

```diff
@@ -406,6 +476,232 @@ impl ForeignClient {
+                    self.id().clone(),
+                    self.src_chain.id(),
+                    format!("{}", e),
+                )
+            })?;
+        consensus_states.sort_by(|a, b| a.height.cmp(&b.height));
+        consensus_states.reverse();
+        Ok(consensus_states)
+    }
+
+    /// Retrieves all consensus heights for this client sorted in reverse
+    /// order.
+    fn consensus_state_heights(&self) -> Result<Vec<Height>, ForeignClientError> {
+        let consensus_state_heights: Vec<Height> = self
+            .consensus_states()?
+            .iter()
+            .filter_map(|cs| Option::from(cs.height))
+            .collect();
+
+        Ok(consensus_state_heights)
+    }
+
+    /// Check for misbehaviour and submit evidence.
+    /// The check starts with and `update_event` emitted by chain B (`dst_chain`) for a client update
+    /// with a header from chain A (`src_chain`). The algorithm goes backwards through the headers
+    /// until it gets to the first misbehaviour.
+    ///
+    /// The following cases are covered:
+    /// 1 - fork:
+    /// Assumes at least one consensus state before the fork point exists.
+    /// Let existing consensus states on chain B be: [Sn,.., Sf, Sf-1, S0] with `Sf-1` being
+    /// the most recent state before fork.
+    /// Chain A is queried for a header `Hf'` at `Sf.height` and if it is different than the `Hf`
+    /// in the event for the client update (the one that has generated `Sf` on chain), then the two
+    /// headers are included in the evidence and submitted.
+    /// Note that in this case the headers are different but have the same height.
+    ///
+    /// 2 - BFT time violation for unavailable header (a.k.a. Future Lunatic Attack or FLA):
+    /// Some header with a height that is higher than the latest
+    /// height on A has been accepted and a consensus state was created on B. Note that this implies
+    /// that the timestamp of this header must be within the `clock_drift` of the client.
+    /// Assume the client on B has been updated with `h2`(not present on/ produced by chain A)
+    /// and it has a timestamp of `t2` that is at most `clock_drift` in the future.
+    /// Then the latest header from A is fetched, let it be `h1`, with a timestamp of `t1`.
+    /// If `t1 >= t2` then evidence of misbehavior is submitted to A.
+    ///
+    /// 3 - BFT time violation for existing headers (TODO):
+    /// Ensure that consensus state times are monotonically increasing with height.
+    ///
+    /// Other notes:
+    /// - the algorithm builds misbehavior at each consensus height, starting with the
+    /// highest height assuming the previous one is trusted. It submits the first constructed
+    /// evidence (the one with the highest height)
+    /// - a lot of the logic here is derived from the behavior of the only implemented client
+    /// (ics07-tendermint) and might not be general enough.
+    ///
+    pub fn handle_misbehaviour(
```

Comment:
Looks good to me, as far as I can tell point 1 and 2 are indeed implemented here and within the light client `build_misbehaviour` method.

We could perhaps make it a bit more explicit where each of the attacks described above are checked, ie. add a reference to the attack number within the inline code comments.

Reply:
I added numbers in `compatible_with`. Let me know if you think i should add in other places.